<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · CHMMAIRRa.jl</title><meta name="title" content="API · CHMMAIRRa.jl"/><meta property="og:title" content="API · CHMMAIRRa.jl"/><meta property="twitter:title" content="API · CHMMAIRRa.jl"/><meta name="description" content="Documentation for CHMMAIRRa.jl."/><meta property="og:description" content="Documentation for CHMMAIRRa.jl."/><meta property="twitter:description" content="Documentation for CHMMAIRRa.jl."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">CHMMAIRRa.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Overview</a></li><li><a class="tocitem" href="../outputs/">Outputs</a></li><li class="is-active"><a class="tocitem" href>API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/MurrellGroup/CHMMAIRRa.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/MurrellGroup/CHMMAIRRa.jl/blob/main/docs/src/api.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="API-Reference"><a class="docs-heading-anchor" href="#API-Reference">API Reference</a><a id="API-Reference-1"></a><a class="docs-heading-anchor-permalink" href="#API-Reference" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CHMMAIRRa.detect_chimeras-Tuple{Tuple{Vector{String}, Vector{String}}, DataFrames.DataFrame}" href="#CHMMAIRRa.detect_chimeras-Tuple{Tuple{Vector{String}, Vector{String}}, DataFrames.DataFrame}"><code>CHMMAIRRa.detect_chimeras</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">detect_chimeras(V_db::Tuple{Vector{String}, Vector{String}}, assignments::DataFrame;
                J_db::Union{Tuple{Vector{String}, Vector{String}}, Nothing} = nothing,
                p_threshold::Float64 = 0.95,
                min_DFR::Int = 1,
                HMM_parameters::Union{Dict, Nothing} = nothing,
                receptor::String = &quot;IG&quot;,

                chimeric_alignments::Bool = false,
                recombfreqplot::Bool = false,
                detailed::Bool = false,
                subsample::Union{Int, Nothing} = nothing,
                deduplicate::Bool = false,


                count_chimeric_segments::Bool = false,
                trim::Int = 5,
                seed::Int = 123,

                mafft::Union{String, Nothing} = nothing,
                align_database::Bool = false,

                quiet::Bool = false,
                disable_internal_dedup::Bool = false
                )::NamedTuple{(:out, :chimeric_alignments, :recombfreqplot),
                            Tuple{DataFrame,
                            Union{Nothing, Tuple{Vector{String}, Vector{String}}},
                            Any}
                            }
                }</code></pre><p>Run CHMMAIRRa on parsed arguments, returning output as a Tuple containing relevant outputs. Output tuple fields:</p><ol><li>out: the input assignments dataframe with chimerism probabilities.</li><li>chimeric<em>alignments: names and sequences for alignments showing which reference sequences recombined at which positions to form detected chimeras. Set to nothing if chimeric</em>alignments is false.</li><li>recombfreqplot: barplot of most frequent recombinations, normalized by their expected frequencies. Set to nothing if recombfreqplot is false.</li></ol><p><strong>Arguments:</strong></p><ul><li><code>V_db::Tuple{Vector{String}, Vector{String}}</code>: Tuple containing (reference names, reference sequences) for the V database.</li><li><code>assignments::DataFrame</code>: VDJ assignments, in MiAIRR format.</li></ul><p><strong>Optional arguments:</strong></p><ul><li><code>J_db::Union{Tuple{Vector{String}, Vector{String}}, Nothing}</code>: Optional tuple containing (reference names, reference sequences) for the J database. If provided, will run on both V and J and the sequence will be chimeric if either V or J is chimeric.</li><li><code>p_threshold::Float64</code>: Posterior threshold of switching templates.</li><li><code>min_DFR::Int</code>: Minimum differences from reference (DFR) threshold to consider a recombination event.</li><li><code>HMM_parameters::Union{Dict, Nothing}</code>: If set, use HMM parameters in this dictionary. Overrides parameter presets from receptor.</li><li><code>receptor::String</code>: Which adaptive immune receptor to run on. Options: IG, TCR.</li><li><code>chimeric_alignments::Bool</code>: Whether to generate and return chimeric sequences and the germline alignments they matched to.</li><li><code>recombfreqplot::Bool</code>: Whether to generate and return a plot of chimerism per recombination event.</li><li><code>detailed::Bool</code>: Whether to include details about the chimeras. Calculate viterbi path for chimeric sequences to add startingpoint, recombinations, recombinations<em>degapped, and max</em>log_prob columns.</li><li><code>subsample::Union{Int, Nothing}</code>: Randomly subsample this number of unique sequences to run chimera detection on.</li><li><code>deduplicate::Bool</code>: Whether to deduplicate input by v<em>sequence</em>alignment column. Note that if subsample is given, subsampling is performed after deduplication.</li><li><code>count_chimeric_segments::Bool</code>: If set, count the number of times chimeric segments appear in nonchimeric sequences. Adds chimera<em>segment</em>counts column.</li><li><code>trim::Int</code>: How many nucleotides to trim off the inner ends of the chimeric segments when searching for matches in nonchimeric sequences.</li><li><code>seed::Int</code>: Seed to use for random subsampling.</li><li><code>mafft::Union{String, Nothing}</code>: Path to MAFFT binary. If not provided, will try to find mafft in system PATH.</li><li><code>align_database::Bool</code>: Whether to align database sequences.</li><li><code>quiet::Bool</code>: If set, hides non-error messages.</li><li><code>disable_internal_dedup::Bool</code>: If set, does not deduplicate sequences internally. This was added for timing benchmark purposes and has no effect on the output.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CHMMAIRRa.jl/blob/2d84cb1d1703b6810f4317a2941b51f8cc0e0788/src/CHMMAIRRa.jl#L363-L423">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CHMMAIRRa.detect_chimeras-Tuple{Vector{String}, Vector{String}, DataFrames.DataFrame, Char}" href="#CHMMAIRRa.detect_chimeras-Tuple{Vector{String}, Vector{String}, DataFrames.DataFrame, Char}"><code>CHMMAIRRa.detect_chimeras</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">detect_chimeras(refnames::Vector{String}, refseqs::Vector{String}, assignments::DataFrame, gene::Char;
                p_threshold::Float64 = 0.95,
                min_DFR::Int = 1,
                HMM_parameters::Union{Dict, Nothing} = nothing,
                receptor::String = &quot;IG&quot;,

                chimeric_alignments::Bool = false,
                recombfreqplot::Bool = false,
                detailed::Bool = false,
                subsample::Union{Int, Nothing} = nothing,
                deduplicate::Bool = false,


                count_chimeric_segments::Bool = false,
                trim::Int = 5,
                seed::Int = 123,

                mafft::Union{String, Nothing} = nothing,
                align_database::Bool = false,

                quiet::Bool = false,
                disable_internal_dedup::Bool = false
                )::NamedTuple{(:out, :chimeric_alignments, :recombfreqplot),
                            Tuple{DataFrame,
                            Union{Nothing, Tuple{Vector{String}, Vector{String}}},
                            Any}
                            }
                }</code></pre><p>Run CHMMAIRRa on parsed arguments, returning output as a Tuple containing relevant outputs. Output tuple fields:</p><ol><li>out: the input assignments dataframe with chimerism probabilities.</li><li>chimeric<em>alignments: names and sequences for alignments showing which reference sequences recombined at which positions to form detected chimeras. Set to nothing if chimeric</em>alignments is false.</li><li>recombfreqplot: barplot of most frequent recombinations, normalized by their expected frequencies. Set to nothing if recombfreqplot is false.</li></ol><p><strong>Arguments:</strong></p><ul><li><code>refnames::Vector{String}</code>: Names of the reference sequences.</li><li><code>refseqs::Vector{String}</code>: Sequences of the reference sequences.</li><li><code>assignments::DataFrame</code>: VDJ assignments, in MiAIRR format.</li><li><code>gene::Char</code>: Which gene to run on. Options: &#39;v&#39;, &#39;j&#39;.</li></ul><p><strong>Optional arguments:</strong></p><ul><li><code>p_threshold::Float64</code>: Posterior threshold of switching templates.</li><li><code>min_DFR::Int</code>: Minimum differences from reference (DFR) threshold to consider a recombination event.</li><li><code>HMM_parameters::Union{Dict, Nothing}</code>: If set, use HMM parameters in this dictionary. Overrides parameter presets from receptor.</li><li><code>receptor::String</code>: Which adaptive immune receptor to run on. Options: IG, TCR.</li><li><code>chimeric_alignments::Bool</code>: Whether to generate and return chimeric sequences and the germline alignments they matched to.</li><li><code>recombfreqplot::Bool</code>: Whether to generate and return a plot of chimerism per recombination event.</li><li><code>detailed::Bool</code>: Whether to include details about the chimeras. Calculate viterbi path for chimeric sequences to add startingpoint, recombinations, recombinations<em>degapped, and max</em>log_prob columns.</li><li><code>subsample::Union{Int, Nothing}</code>: Randomly subsample this number of unique sequences to run chimera detection on.</li><li><code>deduplicate::Bool</code>: Whether to deduplicate input by v<em>sequence</em>alignment column. Note that if subsample is given, subsampling is performed after deduplication.</li><li><code>count_chimeric_segments::Bool</code>: If set, count the number of times chimeric segments appear in nonchimeric sequences. Adds chimera<em>segment</em>counts column.</li><li><code>trim::Int</code>: How many nucleotides to trim off the inner ends of the chimeric segments when searching for matches in nonchimeric sequences.</li><li><code>seed::Int</code>: Seed to use for random subsampling.</li><li><code>mafft::Union{String, Nothing}</code>: Path to MAFFT binary. If not provided, will try to find mafft in system PATH.</li><li><code>align_database::Bool</code>: Whether to align database sequences.</li><li><code>quiet::Bool</code>: If set, hides non-error messages.</li><li><code>disable_internal_dedup::Bool</code>: If set, does not deduplicate sequences internally. This was added for timing benchmark purposes and has no effect on the output.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CHMMAIRRa.jl/blob/2d84cb1d1703b6810f4317a2941b51f8cc0e0788/src/CHMMAIRRa.jl#L539-L599">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CHMMAIRRa.detect_chimeras_from_files-Tuple{String, String, String}" href="#CHMMAIRRa.detect_chimeras_from_files-Tuple{String, String, String}"><code>CHMMAIRRa.detect_chimeras_from_files</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">detect_chimeras_from_files(V_fasta::String, assignments::String, out::String;

        receptor::String = &quot;IG&quot;,
        J_fasta::Union{String, Nothing} = nothing,

        non_chimeric_MiAIRR::Union{String, Nothing} = nothing,
        chimeric_MiAIRR::Union{String, Nothing} = nothing,
        chimeric_alignments::Union{String, Nothing} = nothing,
        recombfreqplot::Union{String, Nothing} = nothing,
        detailed::Bool = false,
        subsample::Union{Int, Nothing} = nothing,
        deduplicate::Bool = false,
        chunk_size::Union{Int, Nothing} = nothing,
       
        p_threshold::Float64 = 0.95,
        min_DFR::Int = 1,
        HMM_parameters::Union{String, Nothing} = nothing,
        mafft::Union{String, Nothing} = nothing,
        align_database::Bool = false,
        count_chimeric_segments::Bool = false,
        trim::Int = 5,
        seed::Int = 123,

        quiet::Bool = false,
        disable_internal_dedup::Bool = false
        )</code></pre><p>Run CHMMAIRRa from input files to writing of output files, taking care of I/O.</p><p><strong>Arguments:</strong></p><ul><li><code>V_fasta::String</code>: V database in fasta format.</li><li><code>assignments::String</code>: VDJ assignments, in MiAIRR format.</li><li><code>out::String</code>: Write CHMMAIRRa output.</li></ul><p><strong>Optional arguments:</strong></p><ul><li><code>receptor::String</code>: Which adaptive immune receptor to run on. Options: IG, TCR.</li><li><code>J_fasta::Union{String, Nothing}</code>: J database in fasta format. If provided, will run on both V and J and the sequence will be chimeric if either V or J is chimeric.</li><li><code>non_chimeric_MiAIRR::Union{String, Nothing}</code>: Write assignments of sequences determined to be non-chimeric.</li><li><code>chimeric_MiAIRR::Union{String, Nothing}</code>: Write assignments of sequences determined to be chimeric.</li><li><code>chimeric_alignments::Union{String, Nothing}</code>: Write fasta file containing chimeric sequences and the germline alignments they matched to.</li><li><code>recombfreqplot::Union{String, Nothing}</code>: Generate a plot of chimerism per recombination event.</li><li><code>detailed::Bool</code>: Include details about the chimeras. Calculate viterbi path for chimeric sequences to add startingpoint, recombinations, recombinations<em>degapped, and max</em>log_prob columns.</li><li><code>subsample::Union{Int, Nothing}</code>: Randomly subsample this number of unique sequences to run chimera detection on.</li><li><code>deduplicate::Bool</code>: Deduplicate input by v<em>sequence</em>alignment column. Note that if subsample is given, subsampling is performed after deduplication.</li><li><code>chunk_size::Union{Int, Nothing}</code>: Read in chunk-size lines of input at a time, write output for that chunk, and move to the next chunk.</li><li><code>p_threshold::Float64</code>: Posterior threshold of switching templates.</li><li><code>min_DFR::Int</code>: Minimum differences from reference (DFR) threshold to consider a recombination event.</li><li><code>HMM_parameters::Union{String, Nothing}</code>: If set, use HMM parameters in this file. Overrides parameter presets from –receptor. See src/IG<em>parameters.tsv or src/TCR</em>parameters.tsv for formatting.</li><li><code>mafft::Union{String, Nothing}</code>: Path to MAFFT binary. If not provided, will try to find mafft in system PATH.</li><li><code>align_database::Bool</code>: Whether to align database sequences.</li><li><code>count_chimeric_segments::Bool</code>: If set, count the number of times chimeric segments appear in nonchimeric sequences. Adds chimera<em>segment</em>counts column.</li><li><code>trim::Int</code>: How many nucleotides to trim off the inner ends of the chimeric segments when searching for matches in nonchimeric sequences.</li><li><code>seed::Int</code>: Seed to use for random subsampling.</li><li><code>quiet::Bool</code>: If set, hides non-error messages.</li><li><code>disable_internal_dedup::Bool</code>: If set, does not deduplicate sequences internally. This was added for timing benchmark purposes and has no effect on the output.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CHMMAIRRa.jl/blob/2d84cb1d1703b6810f4317a2941b51f8cc0e0788/src/CHMMAIRRa.jl#L118-L172">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="CHMMAIRRa.julia_main-Tuple{}" href="#CHMMAIRRa.julia_main-Tuple{}"><code>CHMMAIRRa.julia_main</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">julia_main()::Cint</code></pre><p>Entry point C call for main function for Julia</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/MurrellGroup/CHMMAIRRa.jl/blob/2d84cb1d1703b6810f4317a2941b51f8cc0e0788/src/CHMMAIRRa.jl#L753-L757">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../outputs/">« Outputs</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Wednesday 13 August 2025 13:32">Wednesday 13 August 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
