<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Overview · CHMMAIRRa.jl</title><meta name="title" content="Overview · CHMMAIRRa.jl"/><meta property="og:title" content="Overview · CHMMAIRRa.jl"/><meta property="twitter:title" content="Overview · CHMMAIRRa.jl"/><meta name="description" content="Documentation for CHMMAIRRa.jl."/><meta property="og:description" content="Documentation for CHMMAIRRa.jl."/><meta property="twitter:description" content="Documentation for CHMMAIRRa.jl."/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>CHMMAIRRa.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Overview</a><ul class="internal"><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Usage"><span>Usage</span></a></li><li><a class="tocitem" href="#Arguments"><span>Arguments</span></a></li><li><a class="tocitem" href="#Outputs"><span>Outputs</span></a></li><li><a class="tocitem" href="#Julia-usage"><span>Julia usage</span></a></li><li><a class="tocitem" href="#Citing-CHMMAIRRa.jl"><span>Citing CHMMAIRRa.jl</span></a></li></ul></li><li><a class="tocitem" href="api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Overview</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Overview</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/MurrellGroup/CHMMAIRRa.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/MurrellGroup/CHMMAIRRa.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="CHMMAIRRa.jl"><a class="docs-heading-anchor" href="#CHMMAIRRa.jl">CHMMAIRRa.jl</a><a id="CHMMAIRRa.jl-1"></a><a class="docs-heading-anchor-permalink" href="#CHMMAIRRa.jl" title="Permalink"></a></h1><p><a href="https://opensource.org/license/MIT"><img src="https://img.shields.io/badge/license-MIT-green.svg" alt="MIT license"/></a> <a href="https://codecov.io/gh/MurrellGroup/CHMMAIRRa.jl"><img src="https://codecov.io/gh/MurrellGroup/CHMMAIRRa.jl/branch/main/graph/badge.svg" alt="Coverage"/></a></p><p>CHMMAIRRa.jl is a method for reference-based detection of chimeric adaptive immune receptor repertoires (AIRR) sequences.</p><p>Chimera detection is performed using a Hidden Markov Model (HMM) that models sequences as being generated from a single reference with mutations or from multiple references with mutations. The core HMM functionality is implemented in the <a href="https://github.com/MurrellGroup/CHMMera.jl">CHMMera.jl</a> package, while CHMMAIRRa.jl provides an AIRR-specific wrapper for the HMM.</p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>The tool is primarily distributed as a an executable for Linux and MacOS from the <a href="https://github.com/MurrellGroup/CHMMAIRRa.jl/releases">releases</a> page. Once you have downloaded the executable, you will need to unpack it and add it to your PATH to allow it to be run from any directory in the terminal.</p><pre><code class="nohighlight hljs">unzip CHMMAIRRa_ubuntu-linux_x86-64_v0.0.1.zip # unpack the executable
export PATH=$PATH:$(pwd)/CHMMAIRRa_ubuntu-linux_x86-64_v0.0.1/bin/ # add the executable to your PATH. Place this line in your ~/.bashrc or ~/.zshrc file to make it permanent.</code></pre><p>The tool is also available from the <a href="https://pkg.julialang.org/">Julia package manager</a>, and can be installed with the following command:</p><pre><code class="nohighlight hljs">using Pkg; Pkg.add(&quot;CHMMAIRRa&quot;)</code></pre><p>See the <a href="#Julia-usage">Julia usage</a> section for more details.</p><h2 id="Usage"><a class="docs-heading-anchor" href="#Usage">Usage</a><a id="Usage-1"></a><a class="docs-heading-anchor-permalink" href="#Usage" title="Permalink"></a></h2><p>For chimera detection with a minimal output, run the following command:</p><pre><code class="nohighlight hljs">CHMMAIRRa --receptor IG --V-fasta V.fasta --assignments AIRR.tsv --out CHMMAIRRa_out.tsv</code></pre><p>For chimera detection with a detailed output including recombination path information, run the following command:</p><pre><code class="nohighlight hljs">CHMMAIRRa --receptor IG --V-fasta V.fasta --assignments AIRR.tsv --out CHMMAIRRa_out.tsv --detailed</code></pre><p>If you want the output to include all columns from the input file, use the <code>--non-chimeric-MiAIRR</code> and <code>--chimeric-MiAIRR</code> arguments:</p><pre><code class="nohighlight hljs">CHMMAIRRa --receptor IG --V-fasta V.fasta --assignments AIRR.tsv --out CHMMAIRRa_out.tsv --non-chimeric-MiAIRR nonchimeric.tsv --chimeric-MiAIRR chimeric.tsv</code></pre><h2 id="Arguments"><a class="docs-heading-anchor" href="#Arguments">Arguments</a><a id="Arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Arguments" title="Permalink"></a></h2><h3 id="Mandatory-arguments"><a class="docs-heading-anchor" href="#Mandatory-arguments">Mandatory arguments</a><a id="Mandatory-arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Mandatory-arguments" title="Permalink"></a></h3><ol><li><code>--receptor</code> : The receptor type, either <code>IG</code> or <code>TCR</code>.</li><li><code>--V-fasta</code> : The database V sequences in FASTA format.</li><li><code>--assignments</code> : The VDJ assignments in MiAIRR format. Must have been assigned to the <code>--V-fasta</code> database.</li><li><code>--out</code> : The output file path. See <a href="#chmmairra-output-columns">CHMMAIRRa output columns</a> for more details.</li></ol><h3 id="Optional-arguments"><a class="docs-heading-anchor" href="#Optional-arguments">Optional arguments</a><a id="Optional-arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Optional-arguments" title="Permalink"></a></h3><h4 id="Output-options"><a class="docs-heading-anchor" href="#Output-options">Output options</a><a id="Output-options-1"></a><a class="docs-heading-anchor-permalink" href="#Output-options" title="Permalink"></a></h4><ol><li><code>--detailed</code> : Whether to output detailed information about the recombination path. See <a href="#chmmairra-output-columns">CHMMAIRRa output columns</a> for more details.</li><li><code>--non-chimeric-MiAIRR</code> : The output file path for non-chimeric sequences, including all columns from the input file.</li><li><code>--chimeric-MiAIRR</code> : The output file path for chimeric sequences, including all columns from the input file.</li></ol><h4 id="Extra-outputs"><a class="docs-heading-anchor" href="#Extra-outputs">Extra outputs</a><a id="Extra-outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Extra-outputs" title="Permalink"></a></h4><ol><li><code>--recombfreqplot</code> : A plot of chimerism per recombination event, normalized by allele frequency. Can be useful for identifying missing database sequences.</li><li><code>--chimeric-alignments</code> : Fasta file containing chimeric sequences and the germline alignments they matched to. This also provides a hamming_distances column in the chimeric assignments file.</li></ol><h4 id="Technical-parameters"><a class="docs-heading-anchor" href="#Technical-parameters">Technical parameters</a><a id="Technical-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Technical-parameters" title="Permalink"></a></h4><ol><li><code>--p-threshold</code> : Posterior threshold of switching templates (default: 0.95).</li><li><code>--HMM-parameters</code> : If set, use HMM parameters in this file. Overrides parameter presets from –receptor. See src/IG<em>parameters.tsv or src/TCR</em>parameters.tsv for formatting.</li><li><code>--mafft</code> : The path to the MAFFT executable. Automatically found if not provided.</li><li><code>--align-database</code> : Whether to align the database V sequences. Will be done automatically if database V sequences are not the same length.</li><li><code>--subsample</code> : Randomly subsample this number of unique sequences to run chimera detection on.</li><li><code>--deduplicate</code> : Deduplicate input by v<em>sequence</em>alignment column. Note that if subsample is given, subsampling is performed after deduplication.</li><li><code>--chunk-size</code> : Read in chunk-size lines of input at a time, write output for that chunk, and move to the next chunk. This decreases memory consumption in exchange for increasing runtime.</li><li><code>--count-chimeric-segments</code> : Whether to count the number of times chimeric segments appear in nonchimeric sequences. Adds <code>chimera_segment_counts</code> column.</li><li><code>--trim</code> : How many nucleotides to trim off the inner ends of the chimeric segments when searching for matches in nonchimeric sequences.</li><li><code>--seed</code> : Seed to use for random subsampling.</li><li><code>--quiet</code> : Whether to suppress non-error messages.</li></ol><h2 id="Outputs"><a class="docs-heading-anchor" href="#Outputs">Outputs</a><a id="Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Outputs" title="Permalink"></a></h2><h3 id="Mandatory-columns"><a class="docs-heading-anchor" href="#Mandatory-columns">Mandatory columns</a><a id="Mandatory-columns-1"></a><a class="docs-heading-anchor-permalink" href="#Mandatory-columns" title="Permalink"></a></h3><ul><li><code>threaded</code> : The query V sequence threaded onto the database alignment.</li><li><code>chimera_probability</code> : The probability of the V sequence being a chimera.</li><li><code>chimeric</code> : Whether the V sequence is chimeric.</li></ul><h3 id="Optional-columns-generated-when-using-the-detailed-arg"><a class="docs-heading-anchor" href="#Optional-columns-generated-when-using-the-detailed-arg">Optional columns generated when using the <code>--detailed</code> arg</a><a id="Optional-columns-generated-when-using-the-detailed-arg-1"></a><a class="docs-heading-anchor-permalink" href="#Optional-columns-generated-when-using-the-detailed-arg" title="Permalink"></a></h3><ul><li><code>startingpoint</code> : The nonchimeric starting state of the viterbi path (i.e. the HMM&#39;s v_call).</li><li><code>recombinations</code> : A list of recombination points according to the viterbi path of the query. Format: (v<em>call 1, v</em>call 2, recombination position in threaded query sequence).</li><li><code>recombinations_degapped</code> : Same as recombinations, but with the non-gapped position of the recombination.</li><li><code>pathevaluation</code> : The probability of the second likeliest reference in the HMM. This gives us a measure of confidence for the recombination path itself.</li><li><code>nrecombinations</code> : The number of recombination points in the viterbi path.</li></ul><h3 id="Optional-outputs-generated-when-using-the-count-chimeric-segments-arg"><a class="docs-heading-anchor" href="#Optional-outputs-generated-when-using-the-count-chimeric-segments-arg">Optional outputs generated when using the <code>--count-chimeric-segments</code> arg</a><a id="Optional-outputs-generated-when-using-the-count-chimeric-segments-arg-1"></a><a class="docs-heading-anchor-permalink" href="#Optional-outputs-generated-when-using-the-count-chimeric-segments-arg" title="Permalink"></a></h3><ul><li><code>segment_1_count</code> : The number of times the left chimeric segment appears in nonchimeric sequences.</li><li><code>segment_2_count</code> : The number of times the right chimeric segment appears in nonchimeric sequences.</li></ul><h2 id="Julia-usage"><a class="docs-heading-anchor" href="#Julia-usage">Julia usage</a><a id="Julia-usage-1"></a><a class="docs-heading-anchor-permalink" href="#Julia-usage" title="Permalink"></a></h2><p>CHMMAIRRa.jl&#39;s Julia interface provides two functions: <code>detect_chimeras</code> and <code>detect_chimeras_from_files</code>; the first takes paths to the input files and writes to a file, while the second takes a DataFrame and returns a DataFrame.</p><pre><code class="nohighlight hljs">using CHMMAIRRa

# run from DataFrame
refnames, refseqs = read_fasta(&quot;V.fasta&quot;)
assignments = CSV.read(&quot;MiAIRR.tsv&quot;, DataFrame, delim = &#39;\t&#39;)
chmmairra_output = detect_chimeras(refnames, refseqs, assignments, receptor = &quot;IG&quot;)

# run from file paths
detect_chimeras_from_files(&quot;V.fasta&quot;, &quot;MiAIRR.tsv&quot;, &quot;CHMMAIRRa_out.tsv&quot;, receptor = &quot;IG&quot;)</code></pre><h2 id="Citing-CHMMAIRRa.jl"><a class="docs-heading-anchor" href="#Citing-CHMMAIRRa.jl">Citing CHMMAIRRa.jl</a><a id="Citing-CHMMAIRRa.jl-1"></a><a class="docs-heading-anchor-permalink" href="#Citing-CHMMAIRRa.jl" title="Permalink"></a></h2><p>If you use CHMMAIRRa.jl in your work, please cite our paper:</p><pre><code class="nohighlight hljs"></code></pre></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="api/">API »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Thursday 20 February 2025 16:00">Thursday 20 February 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
